File: src/com/google/code/orion_viewer/SimpleLayoutStrategy.java
package com.google.code.orion_viewer;

/*
 * Orion Viewer is a pdf and djvu viewer for android devices
 *
 * Copyright (C) 2011-2012  Michael Bogdanov
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import android.graphics.Point;
import universe.constellation.orion.viewer.prefs.GlobalOptions;

/**
 * User: mike
 * Date: 15.10.11
 * Time: 13:18
 */
public class SimpleLayoutStrategy implements LayoutStrategy {

    public int viewWidth;

    public int viewHeight;

    private static final int OVERLAP = 15;

    public int VERT_OVERLAP = 3;
    public int HOR_OVERLAP = 3;

    private DocumentWrapper doc;

    private int leftMargin, topMargin, rightMargin, bottomMargin;

    private int zoom;

    private int rotation;

    private int direction;

    private int layout;

    public SimpleLayoutStrategy(DocumentWrapper doc) {
        this.doc = doc;
    }

    public void nextPage(LayoutPosition info) {
        if (getDirection() == 0) {
            if (info.cellX < info.maxX) {
                info.cellX += 1;
            } else if (info.cellY < info.maxY) {
                info.cellX = 0;
                info.cellY += 1;
            } else {
                if (info.pageNumber < doc.getPageCount() - 1) {
                    reset(info, info.pageNumber + 1);
                }
            }
        } else {
            if (info.cellY < info.maxY) {
                info.cellY += 1;
            } else if (info.cellX < info.maxX) {
                info.cellY = 0;
                info.cellX += 1;
            } else {
                if (info.pageNumber < doc.getPageCount() - 1) {
                    reset(info, info.pageNumber + 1);
                }
            }
        }
        Common.d("new cellX = " + info.cellX + " cellY = " + info.cellY);
    }

    public void prevPage(LayoutPosition info) {
        if (getDirection() == 0) {
            if (info.cellX > 0) {
                info.cellX -= 1;
            } else if (info.cellY > 0) {
                info.cellX = info.maxX;
                info.cellY -= 1;
            } else {
                if (info.pageNumber > 0) {
                    reset(info, info.pageNumber - 1);
                    info.cellX = info.maxX;
                    info.cellY = info.maxY;
                }
            }
        } else {
            if (info.cellY > 0) {
                info.cellY -= 1;
            } else if (info.cellX > 0) {
                info.cellY = info.maxY;
                info.cellX -= 1;
            } else {
                if (info.pageNumber > 0) {
                    reset(info, info.pageNumber - 1);
                    info.cellX = info.maxX;
                    info.cellY = info.maxY;
                }
            }
        }

        Common.d("new cellX = " + info.cellX + " maxX = " + info.cellY);
    }
    public boolean changeRotation(int rotation) {
        if (this.rotation != rotation) {
            this.rotation = rotation;
            return true;
        }
        return false;
    }

    public boolean changeOverlapping(int horizontal, int vertical) {
        if (HOR_OVERLAP != horizontal || VERT_OVERLAP != vertical) {
            HOR_OVERLAP = horizontal;
            VERT_OVERLAP = vertical;
            return true;
        }
        return false;
    }

    public void reset(LayoutPosition info, int pageNum) {
        if (doc.getPageCount() - 1 < pageNum) {
            pageNum = doc.getPageCount() - 1;
        }
        if (pageNum < 0) {
            pageNum = 0;
        }

        info.pageNumber = pageNum;
        //original width and height without cropped margins
        PageInfo pinfo = doc.getPageInfo(pageNum);

        info.marginLeft = (int) (leftMargin * pinfo.width * 0.01);
        info.marginTop = (int) (topMargin * pinfo.height * 0.01);

        info.pageWidth = Math.round(pinfo.width * (1 - 0.01f *(leftMargin + rightMargin)));
        info.pageHeight = Math.round(pinfo.height * (1- 0.01f *(topMargin + bottomMargin)));

        info.pieceWidth = rotation == 0 ? viewWidth : viewHeight;
        info.pieceHeight = rotation == 0 ? viewHeight : viewWidth;

        info.screenWidth = viewWidth;
        info.screenHeight = viewHeight;

        //calc zoom
        if (zoom <= 0) {
            //zoom by width
            info.docZoom = 1.0f * info.pieceWidth / info.pageWidth;
        } else {
            info.docZoom = 0.01f * zoom;
        }
        info.marginLeft = (int) (info.docZoom * info.marginLeft);
        info.marginTop = (int) (info.docZoom * info.marginTop);

        //zoomed with and height
        info.pageWidth = (int)(info.docZoom * info.pageWidth);
        info.pageHeight = (int) (info.docZoom * info.pageHeight);

        int hOverlap = info.pieceWidth * HOR_OVERLAP / 100;
        int vOverlap = info.pieceHeight * VERT_OVERLAP / 100;
        //System.out.println("overlap " + hOverlap + " " + vOverlap);
        if (info.pieceHeight != 0 && info.pieceWidth != 0) {
            info.maxX = (info.pageWidth - hOverlap) / (info.pieceWidth - hOverlap) + ((info.pageWidth - hOverlap) % (info.pieceWidth - hOverlap) == 0 ? 0 : 1)  - 1;
            info.maxY = (info.pageHeight - vOverlap) / (info.pieceHeight - vOverlap) + ((info.pageHeight - vOverlap) % (info.pieceHeight - vOverlap) == 0 ?  0: 1) -1;
        }

        info.cellX = 0;
        info.cellY = 0;
    }

    public boolean changeZoom(int zoom) {
        if (this.zoom != zoom) {
            this.zoom = zoom;
            return true;
        }
        return false;
    }

    public boolean changeNavigation(int navigation) {
        if (this.direction != navigation) {
            this.direction = navigation;
            return true;
        }
        return false;
    }

    public boolean changePageLayout(int navigation) {
        if (this.layout != navigation) {
            this.layout = navigation;
            return true;
        }
        return false;
    }

    public int getZoom() {
        return zoom;
    }

    public boolean changeMargins(int leftMargin, int topMargin, int rightMargin, int bottomMargin) {
        if (leftMargin != this.leftMargin || rightMargin != this.rightMargin || topMargin != this.topMargin || bottomMargin != this.bottomMargin) {
            this.leftMargin = leftMargin;
            this.rightMargin = rightMargin;
            this.topMargin = topMargin;
            this.bottomMargin = bottomMargin;
            return true;
        }
        return false;
    }

    public void getMargins(int [] margins) {
        margins[0] = leftMargin;
        margins[1] = rightMargin;
        margins[2] = topMargin;
        margins[3] = bottomMargin;
    }

    public int getRotation() {
        return rotation;
    }


    public void init(LastPageInfo info, GlobalOptions options) {
        changeMargins(info.leftMargin, info.topMargin, info.rightMargin, info.bottomMargin);
        changeRotation(info.rotation);
        changeZoom(info.zoom);
        changeNavigation(info.navigation);
        changePageLayout(info.pageLayout);
        changeOverlapping(options.getHorizontalOverlapping(), options.getVerticalOverlapping());
    }

    public void serialize(LastPageInfo info) {
        info.screenHeight = viewHeight;
        info.screenWidth = viewWidth;

        info.leftMargin = leftMargin;
        info.rightMargin = rightMargin;
        info.topMargin = topMargin;
        info.bottomMargin = bottomMargin;
        info.rotation = rotation;
        info.zoom = zoom;
        info.navigation = direction;
        info.pageLayout = layout;
    }

    public Point convertToPoint(LayoutPosition pos) {
        int layout  = getLayout();
//        int absLeftMargin = (int) (leftMargin * pos.pageWidth * 0.01);
//        int absTopMargin = (int) (topMargin * pos.pageHeight * 0.01);
        int absLeftMargin = (int) pos.marginLeft;
        int absTopMargin = (int) pos.marginTop;
        int hOverlap = pos.pieceWidth * HOR_OVERLAP / 100;
        int vOverlap = pos.pieceHeight * VERT_OVERLAP / 100;

        int x = pos.cellX == 0 || pos.cellX != pos.maxX || layout == 0 ?  (int)(absLeftMargin + pos.cellX * (pos.pieceWidth - hOverlap)) : (int) (absLeftMargin + pos.pageWidth - pos.pieceWidth);
        int y = pos.cellY == 0 || pos.cellY != pos.maxY || layout != 1 ? (int) (absTopMargin + pos.cellY * (pos.pieceHeight - vOverlap)) : (int) (absTopMargin + pos.pageHeight - pos.pieceHeight);

        return new Point(x, y);
    }

    public int getLayout() {
        return layout;
    }

    public int getDirection() {
        return direction;
    }

    public void setDimension(int width, int height) {
        viewWidth = width;
        viewHeight = height;
    }

    public int getLeftMargin() {
        return leftMargin;
    }

    public int getTopMargin() {
        return topMargin;
    }

    public int getRightMargin() {
        return rightMargin;
    }

    public int getBottomMargin() {
        return bottomMargin;
    }
}

File: common/src/commonMain/kotlin/universe/constellation/orion/viewer/layout/SimpleLayoutStrategy.kt
/*
 * Orion Viewer - pdf, djvu, xps and cbz file viewer for android devices
 *
 * Copyright (C) 2011-2017 Michael Bogdanov & Co
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package universe.constellation.orion.viewer.layout

import universe.constellation.orion.viewer.LastPageInfo
import universe.constellation.orion.viewer.PageInfo
import universe.constellation.orion.viewer.PageOptions
import universe.constellation.orion.viewer.PageWalker
import universe.constellation.orion.viewer.geometry.Point
import universe.constellation.orion.viewer.log
import universe.constellation.orion.viewer.walkOrder

class SimpleLayoutStrategy private constructor(
        private val center: Boolean = false
) : LayoutStrategy {

    override var viewWidth = 0
        private set

    override var viewHeight = 0
        private set

    private var VERTICAL_OVERLAP = 3

    private var HORIZONTAL_OVERLAP = 3

    override var margins = CropMargins()
        private set

    override var zoom: Int = 0
        private set

    override var rotation: Int = 0
        private set

    override var walker = PageWalker("default".walkOrder)
        private set

    override var layout: Int = 0
        private set

    override fun nextPage(pos: LayoutPosition): Int {
        if (walker.next(pos, layout)) {
            return 1
        }
        log("new position: $pos")
        return 0
    }

    override fun prevPage(pos: LayoutPosition): Int {
        if (walker.prev(pos, layout)) {
            return -1
        }

        log("new position: $pos")
        return 0
    }

    override fun changeRotation(rotation: Int): Boolean {
        if (this.rotation != rotation) {
            this.rotation = rotation
            return true
        }
        return false
    }

    override fun changeOverlapping(horizontal: Int, vertical: Int): Boolean {
        if (HORIZONTAL_OVERLAP != horizontal || VERTICAL_OVERLAP != vertical) {
            HORIZONTAL_OVERLAP = horizontal
            VERTICAL_OVERLAP = vertical
            return true
        }
        return false
    }

    override fun reset(pos: LayoutPosition, page: PageInfo, forward: Boolean) {
        reset(pos, forward, page, margins.cropMode, zoom, center)
    }

    override fun reset(info: LayoutPosition, forward: Boolean, pageInfo: PageInfo, cropMode: Int, zoom: Int, doCentering: Boolean) {
        info.rotation = rotation
        info.pageNumber = pageInfo.pageNum0

        val pageWidth = pageInfo.width
        val pageHeight = pageInfo.height
        resetMargins(info, pageWidth, pageHeight)

        val isEvenPage = (pageInfo.pageNum0 + 1) % 2 == 0
        val mode = cropMode.toMode

        val autoCrop = pageInfo.autoCrop
        if (autoCrop != null && CropMode.AUTO_MANUAL === mode) {
            appendAutoCropMargins(info, autoCrop)
        }

        if (mode.hasManual()) {
            val leftMargin = if (margins.evenCrop && isEvenPage) margins.evenLeft else margins.left
            val rightMargin = if (margins.evenCrop && isEvenPage) margins.evenRight else margins.right
            appendManualMargins(info, leftMargin, rightMargin)
        }

        if (autoCrop != null && mode.hasAuto() && CropMode.AUTO_MANUAL !== mode) {
            appendAutoCropMargins(info, autoCrop)
        }

        info.x.screenDimension = if (rotation == 0) viewWidth else viewHeight
        info.y.screenDimension = if (rotation == 0) viewHeight else viewWidth

        info.screenWidth = viewWidth
        info.screenHeight = viewHeight

        //set zoom and zoom margins and dimensions
        info.setDocZoom(zoom)

        info.x.marginLeft = (info.docZoom * info.x.marginLeft).toInt()
        info.y.marginLeft = (info.docZoom * info.y.marginLeft).toInt()

        //zoomed with and height
        info.x.pageDimension = (info.docZoom * info.x.pageDimension).toInt()
        info.y.pageDimension = (info.docZoom * info.y.pageDimension).toInt()

        info.x.overlap = info.x.screenDimension * HORIZONTAL_OVERLAP / 100
        info.y.overlap = info.y.screenDimension * VERTICAL_OVERLAP / 100
        //System.out.println("overlap " + hOverlap + " " + vOverlap);

        walker.reset(info, forward, doCentering, layout)
    }

    private fun appendManualMargins(info: LayoutPosition, leftMargin: Int, rightMargin: Int) {
        val pageWidth = info.x.pageDimension
        val pageHeight = info.y.pageDimension

        val xLess = (leftMargin.toDouble() * pageWidth.toDouble() * 0.01).toInt()
        val xMore = (pageWidth.toDouble() * rightMargin.toDouble() * 0.01).toInt()
        val yLess = (margins.top.toDouble() * pageHeight.toDouble() * 0.01).toInt()
        val yMore = (pageHeight.toDouble() * margins.bottom.toDouble() * 0.01).toInt()

        info.x.marginLeft += xLess
        info.x.marginRight += xMore
        info.y.marginLeft += yLess
        info.y.marginRight += yMore

        info.x.pageDimension -= xLess + xMore
        info.y.pageDimension -= yLess + yMore
    }

    private fun appendAutoCropMargins(info: LayoutPosition, autoCrop: AutoCropMargins) {
        info.x.marginLeft += autoCrop.left
        info.x.marginRight += autoCrop.right
        info.y.marginLeft += autoCrop.top
        info.y.marginRight += autoCrop.bottom

        info.x.pageDimension -= autoCrop.left + autoCrop.right
        info.y.pageDimension -= autoCrop.top + autoCrop.bottom
    }

    private fun resetMargins(info: LayoutPosition, pageWidth: Int, pageHeight: Int) {
        info.x.marginLeft = 0
        info.y.marginLeft = 0
        info.x.marginRight = 0
        info.y.marginRight = 0
        info.x.pageDimension = pageWidth
        info.y.pageDimension = pageHeight
    }

    override fun changeZoom(zoom: Int): Boolean {
        if (this.zoom != zoom) {
            this.zoom = zoom
            return true
        }
        return false
    }

    override fun changeWalkOrder(walkOrder: String): Boolean {
        val newWalkOrder = walkOrder.walkOrder
        if (newWalkOrder != walker.order) {
            walker = PageWalker(newWalkOrder)
            return true
        }
        return false
    }

    override fun changePageLayout(pageLayout: Int): Boolean {
        if (this.layout != pageLayout) {
            this.layout = pageLayout
            return true
        }
        return false
    }

    override fun changeCropMargins(cropMargins: CropMargins): Boolean {
        if (cropMargins != this.margins) {
            this.margins = cropMargins
            return true
        }

        return false
    }


    override fun init(info: LastPageInfo, options: PageOptions) {
        changeCropMargins(CropMargins(info.leftMargin, info.rightMargin, info.topMargin, info.bottomMargin, info.leftEvenMargin, info.rightEventMargin, info.enableEvenCropping, info.cropMode))
        changeRotation(info.rotation)
        changeZoom(info.zoom)
        changeWalkOrder(info.walkOrder)
        changePageLayout(info.pageLayout)
        changeOverlapping(options.horizontalOverlapping, options.verticalOverlapping)
    }

    override fun serialize(info: LastPageInfo) {
        info.screenHeight = viewHeight
        info.screenWidth = viewWidth

        info.leftMargin = margins.left
        info.rightMargin = margins.right
        info.topMargin = margins.top
        info.bottomMargin = margins.bottom
        info.leftEvenMargin = margins.evenLeft
        info.rightEventMargin = margins.evenRight
        info.enableEvenCropping = margins.evenCrop
        info.cropMode = margins.cropMode

        info.rotation = rotation
        info.zoom = zoom
        info.walkOrder = walker.order.name
        info.pageLayout = layout
    }

    override fun convertToPoint(pos: LayoutPosition): Point {
        return Point(pos.x.marginLeft + pos.x.offset, pos.y.marginLeft + pos.y.offset)
    }

    override fun setViewSceneDimension(width: Int, height: Int) {
        viewWidth = width
        viewHeight = height
    }

    companion object {

        fun create(): SimpleLayoutStrategy {
            return SimpleLayoutStrategy()
        }
    }
}